### ISSUE TITLE
Classmethod call through protocol-annotated `self` drops `isinstance` narrowing on the bound receiver (regression 0.0.59 → 0.0.60)

### ISSUE THREAD

**Issue by SebbyLaw** [astral-sh/ty#4019]
## Summary

Since 0.0.60, calling a classmethod declared with a protocol-typed `cls` (e.g. `cls: type[P]`) through a `self` that was narrowed via `assert isinstance(self, ...)` reports `invalid-argument-type`. ty computes the bound receiver's class from the **nominal component only** of the narrowed intersection — `self` is `P & Mixin`, but the implicit `cls` argument is reported as `type[Mixin]`, dropping the protocol component, so it no longer satisfies `type[P]`.

## Reproduction

https://play.ty.dev/956c157a-7fef-4738-9918-0e362e70cf47

```python
from typing import TYPE_CHECKING, Protocol


class P(Protocol):
    def __contains__(self, k: str) -> bool: ...


class Helper:
    @classmethod
    def cm(cls: type[P]) -> None: ...


class Mixin(Helper):
    def m(self: P) -> None:
        if TYPE_CHECKING:
            assert isinstance(self, Mixin)
        self.cm()  # error[invalid-argument-type]: Expected `type[P]`, found `type[Mixin]`
```

### Version

ty 0.0.60 (ae554ec78 2026-07-16)

**Comment by charliermarsh** [comment]
I'll take a look at this since I've been iterating on protocols.

**Comment by charliermarsh** [comment]
Okay, I think this is the same as https://github.com/astral-sh/ty/issues/3719. It's possible it'll be fixed by https://github.com/astral-sh/ty/issues/3557?

**Comment by charliermarsh** [comment]
I think `self` gets narrowed to `P & Mixin`, then `self.cm` is found through `Mixin` then `Helper`, and we bind the classmethod through the receiver as if it's just `type[Mixin]`, and that's checked against `type[P]`.

I believe we want the receiver to be `type[P] & type[Mixin]`.

**Comment by carljm** [comment]
It looks like this was actually fixed by astral-sh/ruff#27660.
